From d5092a1171b3fc481130801d3de89b71974cd993 Mon Sep 17 00:00:00 2001
From: pedro <pedro311@gmail.com>
Date: Sun, 19 Jul 2026 19:18:11 +0200
Subject: [PATCH 2/3] ntpd: validate clock step before converting to time_t

The clock offset is derived from peer-provided timestamps.  A malformed
reply can produce a resulting Unix time outside the range of a 32-bit
time_t.  Converting that double to time_t has undefined behavior.

Reject NaN, pre-epoch time, and values outside the signed or unsigned
32-bit time_t range.  If a step is rejected, do not run the step script
or update the clock-discipline state as if it succeeded.

Signed-off-by: pedro311 <pedro311@gmail.com>
---
 busybox/networking/ntpd.c | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/networking/ntpd.c b/networking/ntpd.c
index 49bf0d863..b8464f7da 100644
--- busybox/networking/ntpd.c
+++ busybox/networking/ntpd.c
@@ -1072,7 +1072,7 @@ static void run_script(const char *action, double offset)
 	free(env7);
 }
 
-static NOINLINE void
+static NOINLINE int
 step_time(double offset)
 {
 	llist_t *item;
@@ -1091,8 +1091,22 @@ step_time(double offset)
 	 * a more complex code would be needed.
 	 */
 	dtime = tvc.tv_sec + (1.0e-6 * tvc.tv_usec) + offset;
+	/*
+	 * A peer can make offset large enough that dtime is not representable
+	 * by a 32-bit time_t.  Converting such a double to time_t is undefined.
+	 * ntpd does not support dates before the Unix epoch.
+	 */
+	if (dtime != dtime
+	 || dtime < 0.0
+	 || (sizeof(time_t) == 4
+	     && dtime >= (((time_t)-1 > (time_t)0)
+	                  ? 4294967296.0 : 2147483648.0))
+	) {
+		bb_error_msg("refusing invalid clock step: resulting time %.0f", dtime);
+		return 0;
+	}
 	tvn.tv_sec = (time_t)dtime;
-	tvn.tv_usec = (dtime - tvn.tv_sec) * 1000000;
+	tvn.tv_usec = (dtime - (double)tvn.tv_sec) * 1000000;
 	xsettimeofday(&tvn);
 
 	VERB2 {
@@ -1129,6 +1143,7 @@ step_time(double offset)
 			set_next(pp, RETRY_INTERVAL);
 		}
 	}
+	return 1;
 }
 
 static void clamp_pollexp_and_set_MAXSTRAT(void)
@@ -1570,7 +1585,8 @@ update_local_clock(peer_t *p)
 		 * intervals.
 		 */
 		VERB4 bb_error_msg("stepping time by %+f; poll_exp=MINPOLL", offset);
-		step_time(offset);
+		if (!step_time(offset))
+			return 0;
 		if (option_mask32 & OPT_q) {
 			/* We were only asked to set time once. Done. */
 			exit(0);
-- 
2.39.5

